db-engine/src/main/java/com/splicemachine/db/iapi/stats/FakeColumnStatisticsImpl.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2012 - 2019 Splice Machine, Inc.
+ *
+ * This file is part of Splice Machine.
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.db.iapi.stats;
+
+import com.splicemachine.db.iapi.types.DataValueDescriptor;
+import org.apache.log4j.Logger;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+
+/**
+ * Created by yxia on 11/7/19.
+ */
+public class FakeColumnStatisticsImpl extends ColumnStatisticsImpl implements Externalizable {
+    private static Logger LOG=Logger.getLogger(FakeColumnStatisticsImpl.class);
+    private long nullCount;
+    private long totalCount;
+    private long cardinality;
+    private long rpv;
+
+    public FakeColumnStatisticsImpl() {
+    }
+
+    public FakeColumnStatisticsImpl(DataValueDescriptor dvd,
+                             long nullCount,
+                             long totalCount,
+                             long cardinality) {
+        this.dvd = dvd;
+        this.nullCount = nullCount;
+        this.totalCount = totalCount;
+        if (totalCount < nullCount)
+            this.totalCount = nullCount;
+        this.cardinality = cardinality;
+        if (cardinality > this.totalCount - nullCount + 1)
+            this.cardinality = this.totalCount - nullCount + 1;
+        rpv = (long) (((double)(this.totalCount - nullCount))/cardinality);
```

Comment:
Should catch a zero or negative cardinality and handle it somehow (use totalCount or 1 as the default?)

Reply:
Done. I've added the following logic in StatisticsAdmin.FAKE_COLUMN_STATISTICS to check the eligible values for cardinality:
```
            if (rpv > totalCount - nullCount || rpv < 1)
                throw ErrorState.LANG_INVALID_FAKE_STATS.newException("column", "rows per value shouldn't be less than 1 or larger than the total number of not-null count : " + (totalCount - nullCount));

            long cardinality = (long)(((double)(totalCount- nullCount))/rpv);
```